durable-storage/src/persistence_layer.rs

```diff
@@ -110,10 +136,17 @@ pub struct PersistenceLayer {
     db_instance: ManuallyDrop<rocksdb::DB>,
 
     /// What mode was the [`PersistenceLayer`] opened in.
-    mode: Mode,
+    mode: Mode<'a>,
 }
 
-impl PersistenceLayer {
+impl<'a> PersistenceLayer<'a> {
+    /// Creates a checkpoint of the current database at the given `path`.
+    fn checkpoint_at(&self, path: &PathBuf) -> Result<(), Error> {
```

Comment:
`&PathBuf` requires a `PathBuf`, whereas `&Path` works for all sorts of paths. Even better would be `impl AsRef<Path>`, like `create_checkpoint` requires.

```suggestion
    fn checkpoint_at(&self, path: &Path) -> Result<(), Error> {
```